taichi/codegen/codegen_llvm_quant.cpp

```diff
@@ -145,51 +151,153 @@ void CodeGenLLVM::visit(BitStructStoreStmt *stmt) {
+      auto cft = dtype->as<CustomFloatType>();
+      llvm::Value *digit_bits = nullptr;
+      // Extract exponent and digits from compute type (assumed to be f32 for
+      // now).
+      TI_ASSERT(cft->get_compute_type()->is_primitive(PrimitiveTypeID::f32));
+
+      // f32 = 1 sign bit + 8 exponent bits + 23 fraction bits
+
+      auto f32_bits =
+          builder->CreateBitCast(val, llvm::Type::getInt32Ty(*llvm_context));
+      // Rounding to nearest here. Note that if the digits overflows then the
+      // carry-on will contribute to the exponent, which is desired.
+      if (cft->get_digit_bits() < 23) {
+        f32_bits = builder->CreateAdd(
+            f32_bits, tlctx->get_constant(1 << (22 - cft->get_digit_bits())));
+      }
+
+      auto exponent_bits = builder->CreateAShr(f32_bits, 23);
+      exponent_bits =
+          builder->CreateAnd(exponent_bits, tlctx->get_constant((1 << 8) - 1));
+      auto value_bits = builder->CreateAShr(
+          f32_bits, tlctx->get_constant(23 - cft->get_digit_bits()));
+
+      digit_bits = builder->CreateAnd(
+          value_bits, tlctx->get_constant((1 << (cft->get_digit_bits())) - 1));
+
+      if (cft->get_is_signed()) {
+        // extract the sign bit
+        auto sign_bit =
+            builder->CreateAnd(f32_bits, tlctx->get_constant(0x80000000u));
+        // insert the sign bit to digit bits
+        digit_bits = builder->CreateOr(
+            digit_bits,
+            builder->CreateLShr(sign_bit, 31 - cft->get_digit_bits()));
+      }
+
+      auto digits_snode = ch.get();
+      auto exponent_snode = digits_snode->exp_snode;
+
+      auto exponent_offset = get_exponent_offset(exponent_bits, cft);
+      exponent_bits = builder->CreateSub(exponent_bits, exponent_offset);
+      exponent_bits =
+          create_call("max_i32", {exponent_bits, tlctx->get_constant(0)});
+
+      // Compute the bit pointer of the exponent bits.
+      TI_ASSERT(digits_snode->parent == exponent_snode->parent);
+
+      val = builder->CreateBitCast(exponent_bits,
+                                   llvm_type(bit_struct_physical_type));
+      val = builder->CreateShl(val, exponent_snode->bit_offset);
+
+      if (bit_struct_val == nullptr) {
+        bit_struct_val = val;
+      } else {
+        bit_struct_val = builder->CreateOr(bit_struct_val, val);
+      }
+      // Here we implement flush to zero (FTZ): if exponent is zero, we force
+      // the digits to be zero.
+      // TODO: it seems that this can be more efficiently implemented using a
+      // bit_and.
+      auto exp_non_zero =
+          builder->CreateICmp(llvm::CmpInst::Predicate::ICMP_NE, exponent_bits,
+                              tlctx->get_constant(0));
+      val = builder->CreateSelect(exp_non_zero, digit_bits,
+                                  tlctx->get_constant(0));
+      val = builder->CreateBitCast(val, llvm_type(bit_struct_physical_type));
+      val = builder->CreateShl(val, digits_snode->bit_offset);
```

Comment:
This part seems to overlap with `visit(GlobalStoreStmt *stmt)` - maybe we can extract the common parts into a function?

Reply:
Sorry I didn't realize your intention is to remove the custom float store part in `GlobalStoreStmt` - how about we remove the code in `GlobalStore` in this PR, and insert a `TI_ERROR` there if any `bit_struct` store is dispatched to `GlobalStore` instead of `BitStructStore`?